### ISSUE TITLE
design an API

### ISSUE THREAD

**Issue by duckinator** [how-is/how_is#115]
Instead of haphazardly adding things, actually _design_ an API.

Needs to do the following:

* report from repo+format
  * Given a repository and a format, generate a report object.
  * Given a repository and a filename (which the format is inferred from), generate a report file.
* report from config
  * Given a config object, generate a set of report objects.
  * Given a config object, generate a set of report files.
  * Given a config file, generate a set of report objects
  * Given a config file, generate a set of report files.
* report from other report
  * Given a JSON report object, generate another report object.
  * Given a JSON report object, generate another report file.
  * Given a JSON report file, generate a report object.
  * Given a JSON report file, generate a report file.

**Comment by duckinator** [comment]
IDEA:

Something along the lines of `.from_X(...).to_Y(...)`.

E.g.,

```ruby
# from repo+format
HowIs.from_???(repo).to_object(format)
HowIs.from_???(repo).to_file(filename)

# from config
HowIs.from_config(config).to_object(format)
HowIs.from_config(config).to_file(filename)
HowIs.from_config_file(config).to_object(format)
HowIs.from_config_file(config).to_file(filename)

# from other report
HowIs.from_report(report).to_object(format)
HowIs.from_report(report).to_file(filename)
HowIs.from_report_file(report).to_object(format)
HowIs.from_report_file(report).to_file(filename)
```

**Comment by indirect** [comment]
My personal preference would be an API something like this:

```ruby
# create a report
report = HowIs.new("bundler/bundler")

# export a report to json
report.to_json

# export a report to html
report.to_html

# write out report json to a file
File.write("report.json", report.to_json)

# import previously saved json into a report object
report = HowIs.from_json(File.read("report.json"))

# write out a report HTML to a file
File.write("report.html", report.to_html)
```

I _think_ that covers all the use cases you mentioned. Am I missing any?

**Comment by duckinator** [comment]
That doesn't include generating a report from a Ruby Hash, which was I included more for consistency than anything else.

It also doesn't include generating reports from a Ruby config Hash/file, which is important.

I think I'm going to go with

```ruby
# Generate a new report.
report = HowIs.new('bundler/bundler')

# Generate a group of reports from a config Hash
reports = HowIs.from_config(config)
```

and everything else being what you suggested.

**Comment by duckinator** [comment]
So:

```
# create a report
report = HowIs.new("bundler/bundler")

# convert a report to json string
report.to_json

# convert a report to an html string
report.to_html

# write out report json to a file
File.write("report.json", report.to_json)

# import previously saved json into a report object
report = HowIs.from_json(File.read("report.json"))

# write out a report HTML to a file
File.write("report.html", report.to_html)

# generate a series of reports from a config Hash
reports = HowIs.from_config(config)
```

**Comment by duckinator** [comment]
I'm also trying to figure out how I should do generating a full HTML file vs generating an HTML partial.

I think I'm just going to have `.to_html_partial` and `.to_html` (which would be implemented using `.to_html_partial`). My reasoning there is that `.to_json` is a full file, so for consistency `.to_html` being a full file makes more sense to me.

**Comment by duckinator** [comment]
I believe I've updated all of the specs related to the API change. Now to begin implementing the changes!

**Comment by duckinator** [comment]
Closed by #117.
